File: CS/DXGridCRUDoperations/Form1.cs
// Developer Express Code Central Example:
// How to implement simple copy/paste operations in the GridView from a toolbar menu
// 
// This example illustrates how to implement simple copy/paste/delete operations in
// the GridView.
// You can find more in the article
// http://www.devexpress.com/scid=A1266
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E4528

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.Skins;
using DevExpress.LookAndFeel;
using DevExpress.UserSkins;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Grid;


namespace DXGridCRUDoperations
{
    public partial class Form1 : XtraForm
    {
        public Form1()
        {
            InitializeComponent();
            InitGrid();

        }
        BindingList<Person> gridDataList = new BindingList<Person>();
        void InitGrid()
        {
            gridDataList.Add(new Person("John", "Smith"));
            gridDataList.Add(new Person("Gabriel", "Smith"));
            gridDataList.Add(new Person("Ashley", "Smith", "some comment"));
            gridDataList.Add(new Person("Adrian", "Smith", "some comment"));
            gridDataList.Add(new Person("Gabriella", "Smith", "some comment"));
            gridControl.DataSource = gridDataList;
            gridView1.InitNewRow += new DevExpress.XtraGrid.Views.Grid.InitNewRowEventHandler(gridView1_InitNewRow);
        }

        void gridView1_InitNewRow(object sender, DevExpress.XtraGrid.Views.Grid.InitNewRowEventArgs e)
        {
            GridView view = sender as GridView;

            if (view !=null && personToEdit != null)
            {               
                view.SetRowCellValue(e.RowHandle, view.Columns["FirstName"],personToEdit.FirstName);
                view.SetRowCellValue(e.RowHandle, view.Columns["SecondName"], personToEdit.SecondName);
                view.SetRowCellValue(e.RowHandle, view.Columns["Info"], personToEdit.Info); 
            }           
        }
        private EditForm f1;
        private Person personToEdit;
        BindingList<Person> personList = new BindingList<Person>();
        private void iAdd_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
            personToEdit = new Person();
            EditPerson(personToEdit, "NewPerson", CloseNewPersonHandler);
        }
        private void EditPerson(Person person, string windowTitle, FormClosingEventHandler closedDelegate)
        {
            f1 = new EditForm(person, false) { Text = windowTitle };
            f1.FormClosing += closedDelegate;
            f1.ShowDialog();
        }
        private void CloseNewPersonHandler(object sender, FormClosingEventArgs e)
        {

            if (((EditForm)sender).DialogResult == DialogResult.OK)
            {
                try
                {                    
                    // personList.Add(personToEdit); //add via Datasource                    
                   //((IList)gridView1.DataSource).Add(personToEdit);  
                    gridView1.AddNewRow();
                }
                catch (Exception ex)
                {
                    HandleExcepton(ex);
                }
                SetNewFocus();
            }
            personToEdit = null;
        }
        private void SetNewFocus()
        {
            for (int i = 0; i < gridView1.DataRowCount; i++)
            {
                if (personToEdit.FirstName == gridView1.GetRowCellValue(i, gridView1.Columns["FirstName"]).ToString())
                {
                    gridView1.FocusedRowHandle = i;
                    break;
                }
            }
        }
        private void HandleExcepton(Exception ex)
        {
            MessageBox.Show(ex.Message);
        }
        private void iCopy_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
            if (gridView1.FocusedRowHandle > -1)
            {
                string s = String.Empty;
                for (int i = 0; i < gridView1.Columns.Count; i++)
                {
                    s += gridView1.GetRowCellDisplayText(gridView1.FocusedRowHandle, gridView1.Columns[i]) + "\n";
                }
                Clipboard.SetDataObject(s);
            }
            else
            {
                MessageBox.Show("Row is not selected");
            }
        }
        private void iPaste_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
            if (gridView1.FocusedRowHandle > -1)
            {
                IDataObject data = Clipboard.GetDataObject();
                if (data.GetDataPresent(DataFormats.Text) && gridView1.FocusedRowHandle > -1)
                {
                    string s = data.GetData(DataFormats.Text).ToString();
                    string[] a = s.Split('\n');
                    int i = 0;
                    foreach (GridColumn item in gridView1.Columns)
                    {
                        gridView1.SetRowCellValue(gridView1.FocusedRowHandle, item, a[i]);
                        i++;
                    }
                }
            }
            else
            {
                MessageBox.Show("Row is not selected");
            }
        }
        private void iInsert_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
            personToEdit = new Person();
            InsertPerson(personToEdit, "InsertForm", InsertPersonHandler);
        }
        private void InsertPerson(Person person, string windowTitle, FormClosingEventHandler closedDelegate)
        {
            f1 = new EditForm(person, true) { Text = windowTitle };
            f1.FormClosing += closedDelegate;
            f1.ShowDialog();
        }
        private void InsertPersonHandler(object sender, FormClosingEventArgs e)
        {
            EditForm form = sender as EditForm;
            if (form != null)
            {
                if (form.DialogResult == DialogResult.OK)
                {
                    try
                    {
                        if (!String.IsNullOrWhiteSpace(form.NewPosition))
                        {
                            int pos = Convert.ToInt32(form.NewPosition);
                            if (pos > personList.Count - 1)
                            {

                                personList.Add(personToEdit);
                            }
                            else
                            {
                                
                                personList.Insert(pos, personToEdit);
                            }
                        }
                        else
                        {
                            MessageBox.Show("Incorrect position value");
                        }

                    }
                    catch (Exception ex)
                    {
                        HandleExcepton(ex);
                    }
                }
                SetNewFocus();
            }
            personToEdit = null;
        }
        private void iDelete_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
            DeletePerson(gridView1.FocusedRowHandle);
        }
        public Person GetCurrentPerson()
        {
            return gridView1.GetRow(gridView1.FocusedRowHandle) as Person;
        }
        private void DeletePerson(int focusedRowHandle)
        {
            if (focusedRowHandle < 0)
            {
                MessageBox.Show("You didn't select a person to delete");
                return;
            }
            if (MessageBox.Show("Do you really want to delete the selected person?", "Delete Person", MessageBoxButtons.OKCancel) != DialogResult.OK)
            {
                return;
            }
            try
            {
                //Deleting from datasource
                //personToEdit = GetCurrentPerson();
                //personList.Remove(personToEdit);
                //Deleting from View
                gridView1.DeleteRow(gridView1.FocusedRowHandle);
            }
            catch (Exception ex)
            {
                HandleExcepton(ex);
            }
            gridView1.FocusedRowHandle = focusedRowHandle;
            personToEdit = null;
        }

        private void barButtonItem1_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
            if (gridView1.IsMultiSelect && gridView1.SelectedRowsCount > 0)
            {
                gridView1.DeleteSelectedRows();
            }
        }

    }
}

File: VB/DXGridCRUDoperations/Form1.vb
' Developer Express Code Central Example:
' How to implement simple copy/paste operations in the GridView from a toolbar menu
' 
' This example illustrates how to implement simple copy/paste/delete operations in
' the GridView.
' You can find more in the article
' http://www.devexpress.com/scid=A1266
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=E4528

Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.Skins
Imports DevExpress.LookAndFeel
Imports DevExpress.UserSkins
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Grid


Namespace DXGridCRUDoperations
	Partial Public Class Form1
		Inherits XtraForm
		Public Sub New()
			InitializeComponent()
			InitGrid()

		End Sub
		Private gridDataList As New BindingList(Of Person)()
		Private Sub InitGrid()
			gridDataList.Add(New Person("John", "Smith"))
			gridDataList.Add(New Person("Gabriel", "Smith"))
			gridDataList.Add(New Person("Ashley", "Smith", "some comment"))
			gridDataList.Add(New Person("Adrian", "Smith", "some comment"))
			gridDataList.Add(New Person("Gabriella", "Smith", "some comment"))
			gridControl.DataSource = gridDataList
			AddHandler gridView1.InitNewRow, AddressOf gridView1_InitNewRow
		End Sub

		Private Sub gridView1_InitNewRow(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Grid.InitNewRowEventArgs)
			Dim view As GridView = TryCast(sender, GridView)

			If view IsNot Nothing AndAlso personToEdit IsNot Nothing Then
				view.SetRowCellValue(e.RowHandle, view.Columns("FirstName"),personToEdit.FirstName)
				view.SetRowCellValue(e.RowHandle, view.Columns("SecondName"), personToEdit.SecondName)
				view.SetRowCellValue(e.RowHandle, view.Columns("Info"), personToEdit.Info)
			End If
		End Sub
		Private f1 As EditForm
		Private personToEdit As Person
		Private personList As New BindingList(Of Person)()
		Private Sub iAdd_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles iAdd.ItemClick
			personToEdit = New Person()
			EditPerson(personToEdit, "NewPerson", AddressOf CloseNewPersonHandler)
		End Sub
		Private Sub EditPerson(ByVal person As Person, ByVal windowTitle As String, ByVal closedDelegate As FormClosingEventHandler)
			f1 = New EditForm(person, False) With {.Text = windowTitle}
			AddHandler f1.FormClosing, closedDelegate
			f1.ShowDialog()
		End Sub
		Private Sub CloseNewPersonHandler(ByVal sender As Object, ByVal e As FormClosingEventArgs)

			If (CType(sender, EditForm)).DialogResult = System.Windows.Forms.DialogResult.OK Then
				Try
					' personList.Add(personToEdit); //add via Datasource                    
				   '((IList)gridView1.DataSource).Add(personToEdit);  
					gridView1.AddNewRow()
				Catch ex As Exception
					HandleExcepton(ex)
				End Try
				SetNewFocus()
			End If
			personToEdit = Nothing
		End Sub
		Private Sub SetNewFocus()
			For i As Integer = 0 To gridView1.DataRowCount - 1
				If personToEdit.FirstName = gridView1.GetRowCellValue(i, gridView1.Columns("FirstName")).ToString() Then
					gridView1.FocusedRowHandle = i
					Exit For
				End If
			Next i
		End Sub
		Private Sub HandleExcepton(ByVal ex As Exception)
			MessageBox.Show(ex.Message)
		End Sub
		Private Sub iCopy_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles iCopy.ItemClick
			If gridView1.FocusedRowHandle > -1 Then
				Dim s As String = String.Empty
				For i As Integer = 0 To gridView1.Columns.Count - 1
					s &= gridView1.GetRowCellDisplayText(gridView1.FocusedRowHandle, gridView1.Columns(i)) + Constants.vbLf
				Next i
				Clipboard.SetDataObject(s)
			Else
				MessageBox.Show("Row is not selected")
			End If
		End Sub
		Private Sub iPaste_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles iPaste.ItemClick
			If gridView1.FocusedRowHandle > -1 Then
				Dim data As IDataObject = Clipboard.GetDataObject()
				If data.GetDataPresent(DataFormats.Text) AndAlso gridView1.FocusedRowHandle > -1 Then
					Dim s As String = data.GetData(DataFormats.Text).ToString()
					Dim a() As String = s.Split(ControlChars.Lf)
					Dim i As Integer = 0
					For Each item As GridColumn In gridView1.Columns
						gridView1.SetRowCellValue(gridView1.FocusedRowHandle, item, a(i))
						i += 1
					Next item
				End If
			Else
				MessageBox.Show("Row is not selected")
			End If
		End Sub
		Private Sub iInsert_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles iInsert.ItemClick
			personToEdit = New Person()
			InsertPerson(personToEdit, "InsertForm", AddressOf InsertPersonHandler)
		End Sub
		Private Sub InsertPerson(ByVal person As Person, ByVal windowTitle As String, ByVal closedDelegate As FormClosingEventHandler)
			f1 = New EditForm(person, True) With {.Text = windowTitle}
			AddHandler f1.FormClosing, closedDelegate
			f1.ShowDialog()
		End Sub
		Private Sub InsertPersonHandler(ByVal sender As Object, ByVal e As FormClosingEventArgs)
			Dim form As EditForm = TryCast(sender, EditForm)
			If form IsNot Nothing Then
				If form.DialogResult = System.Windows.Forms.DialogResult.OK Then
					Try
						If (Not String.IsNullOrWhiteSpace(form.NewPosition)) Then
							Dim pos As Integer = Convert.ToInt32(form.NewPosition)
							If pos > personList.Count - 1 Then

								personList.Add(personToEdit)
							Else

								personList.Insert(pos, personToEdit)
							End If
						Else
							MessageBox.Show("Incorrect position value")
						End If

					Catch ex As Exception
						HandleExcepton(ex)
					End Try
				End If
				SetNewFocus()
			End If
			personToEdit = Nothing
		End Sub
		Private Sub iDelete_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles iDelete.ItemClick
			DeletePerson(gridView1.FocusedRowHandle)
		End Sub
		Public Function GetCurrentPerson() As Person
			Return TryCast(gridView1.GetRow(gridView1.FocusedRowHandle), Person)
		End Function
		Private Sub DeletePerson(ByVal focusedRowHandle As Integer)
			If focusedRowHandle < 0 Then
				MessageBox.Show("You didn't select a person to delete")
				Return
			End If
			If MessageBox.Show("Do you really want to delete the selected person?", "Delete Person", MessageBoxButtons.OKCancel) <> System.Windows.Forms.DialogResult.OK Then
				Return
			End If
			Try
				'Deleting from datasource
				'personToEdit = GetCurrentPerson();
				'personList.Remove(personToEdit);
				'Deleting from View
				gridView1.DeleteRow(gridView1.FocusedRowHandle)
			Catch ex As Exception
				HandleExcepton(ex)
			End Try
			gridView1.FocusedRowHandle = focusedRowHandle
			personToEdit = Nothing
		End Sub

		Private Sub barButtonItem1_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles barButtonItem1.ItemClick
			If gridView1.IsMultiSelect AndAlso gridView1.SelectedRowsCount > 0 Then
				gridView1.DeleteSelectedRows()
			End If
		End Sub

	End Class
End Namespace